csharp/ql/src/semmlecode.csharp.dbscheme

```diff
@@ -479,6 +479,34 @@ type_mention_location(
   unique int id: @type_mention ref,
   int loc: @location ref);
 
+@has_type_annotation = @assignable | @type_parameter | @callable | @array_type | @expr | @delegate_type;
+
+/**
+ * A direct annotation on an entity, for example `string? x;`.
+ *
+ * Annotations:
+ * 0 = reftype nullability is not applicable
+ * 1 = reftype nullability disabled
+ * 2 = reftype is not annotated "!"
+ * 3 = reftype is annotated "?"
+ * 4 = readonly ref type / in parameter
+ * 5 = ref type parameter, return or local variable
+ * 6 = out parameter
+ *
+ * Note that the annotation depends on the element it annotates.
+ * @assignable: The annotation is on the type of the assignable, for example the variable type.
+ * @type_parameter: The annotation is on the reftype constraint
+ * @callable: The annotation is on the return type
+ * @array_type: The annotation is on the element type
+ */
+type_annotation(int id: @has_type_annotation ref, int annotation: int ref);
```

Comment:
If we also add `in`, `this`, and `params` as type annotations, we can get rid of the `mode` column from the `params` relation; what do you think?

Reply:
I'm not sure. We already handle `in` (it's the same as `ref readonly`), and `this` and `params` don't feel like types to me. I'd prefer to leave this as-is for now.